src/outpack/location_pull.py

```diff
@@ -0,0 +1,463 @@
+    filename = path_metadata / packet_id
+    with open(filename, "w") as f:
+        f.writelines(metadata)
+
+
+def _get_remove_location_hint(location_name):
+    return (
+        f'Probably all you can do at this point is '
+        f'remove this location from your configuration '
+        f'by running '
+        f'outpack_location_remove("{location_name}")'
+    )
+
+
+def _validate_hashes(driver, location_name, packets: List[PacketLocation]):
+    mismatched_hashes = set()
+    known_there = driver.list()
+    for packet in packets:
+        if known_there.get(packet.packet) is not None:
+            hash_there = known_there[packet.packet].hash
+            hash_here = packet.hash
+            if hash_there != hash_here:
+                mismatched_hashes.add(packet.packet)
+
+    if len(mismatched_hashes) > 0:
+        id_text = "', '".join(mismatched_hashes)
+        msg = (
+            f"Location '{location_name}' has conflicting metadata\n"
+            f"This is really bad news. We have been offered metadata "
+            f"from '{location_name}' that has a different hash to "
+            f"metadata that we have already imported from other "
+            f"locations. I'm not going to import this new metadata, "
+            f"but there's no guarantee that the older metadata is "
+            f"actually what you want!\nConflicts for: '{id_text}'\n"
+            f"We would be interested in this case, please let us know\n"
+            f"{_get_remove_location_hint(location_name)}"
+        )
+        raise Exception(msg)
+
+
+def _mark_all_known(driver, root, location_name):
+    try:
+        known_here = root.index.location(location_name)
+    except KeyError:
+        known_here = {}
+
+    known_there = driver.list()
+    for packet_id in known_there:
+        if packet_id not in known_here.keys():
+            mark_known(
+                root,
+                packet_id,
+                location_name,
+                known_there[packet_id].hash,
+                known_there[packet_id].time,
+            )
+
+
+def outpack_location_pull_packet(
+    ids: Union[str, List[str]],
+    options=None,
+    recursive=None,
+    root=None,
+    *,
+    locate=True,
+):
+    root = root_open(root, locate=locate)
+    options = SearchOptions(options)
+    if isinstance(ids, str):
+        ids = [ids]
+    plan = _location_build_pull_plan(ids, options.location, recursive, root)
+
+    ## Warn people of extra pulls and skips
+    if plan.info.n_extra > 0:
+        print(
+            f"Also pulling {plan.info.n_extra} "
+            f"{pl(plan.info.n_extra, 'packet')}, which are "
+            f"dependencies of those requested"
+        )
+    if plan.info.n_skip > 0:
+        print(
+            f"Skipping {plan.info.n_skip} of {plan.info.n_total} "
+            f"{pl(plan.info.n_total, 'packet')} which are already "
+            f"unpacked"
+        )
+
+    store, cleanup = _location_pull_files(plan.files, root)
```

Comment:
This should be
```python
store, cleanup = _location_pull_files(plan.files, root)
try:
  ...
finally:
  cleanup()
```

in order to make this robust to exceptions. Or make `_location_pull_files` a context manager.